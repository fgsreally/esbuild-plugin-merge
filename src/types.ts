import { BuildOptions, OnLoadArgs, OnLoadOptions, OnLoadResult, OnResolveArgs, OnResolveResult, OnStartResult, Plugin, PluginBuild } from "esbuild"

export interface PluginCb<P, R> {
    filter: RegExp
    callback: (params: P,args?:any) => R
}


export interface ProxyBuilder {
    initialOptions: BuildOptions,
    resolveCb: PluginCb<OnResolveArgs, OnResolveResult | null | void | Promise<OnResolveResult | null | undefined>>[],
    startCb: (() => OnStartResult | void)[],
    endCb: (() => void)[],
    loadCb: PluginCb<OnLoadArgs, OnLoadResult | void | null | Promise<OnLoadResult | null | undefined>>[],
    transformCb: PluginCb<OnLoadResult, void>[],
    updateCb:((id:string,ctx:{event:'update'|'delete'})=>void)[],

    esbuild: PluginBuild["esbuild"]
    onResolve: PluginBuild["onResolve"]
    onLoad: PluginBuild["onLoad"]
    onTransForm: (options: OnLoadOptions, cb: (param: OnLoadResult,args:OnLoadOptions) => void) => void
    onUpdate:(cb:(id:string,ctx:{event:'update'|'delete'})=>void)=>void
    plugins: string[]
}

export type ProPlugin = {
    name: string;
    setup: (build: ProxyBuilder) => (void | Promise<void>);
    enforce?: 'pre' | 'post'
}