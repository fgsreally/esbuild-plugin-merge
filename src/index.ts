import { OnLoadResult, Plugin, PluginBuild } from 'esbuild'
import { ProPlugin, ProxyBuilder } from './types'
import chokidar from 'chokidar'
import fs from 'fs'
import { resolve } from 'path'

type ResolvedPlugin = Plugin & { enforce?: 'pre' | 'post' } | ProPlugin
export function pre(plugin: ResolvedPlugin) {
    plugin.enforce = 'pre'
    return plugin as Plugin
}
export function post(plugin: ResolvedPlugin) {
    plugin.enforce = 'post'
    return plugin as Plugin

}

const watchListRecord: Record<string, chokidar.FSWatcher> = {}
const watchList: Set<string> = new Set()


export const proxyBuilder: ProxyBuilder = {
    onResolve: (options, cb) => {
        proxyBuilder.resolveCb.push({
            filter: options.filter,
            callback: cb
        })
    },
    onLoad: (options, cb) => {
        proxyBuilder.loadCb.push({
            filter: options.filter,
            callback: cb
        })
    },
    onTransForm: (options, cb) => {
        proxyBuilder.transformCb.push({
            filter: options.filter,
            callback: cb
        })
    },
    onUpdate: (cb) => {
        proxyBuilder.updateCb.push(cb)
    },
    onStart:(cb)=>{ proxyBuilder.startCb.push(cb)},
    onEnd:(cb)=>{ proxyBuilder.endCb.push(cb)},

    esbuild: null as any,
    initialOptions: null as any,
    resolveCb: [],
    startCb: [],
    endCb: [],
    loadCb: [],
    transformCb: [],
    updateCb:[],
    plugins: []
}


function sortPlugins(plugins: ResolvedPlugin[],) {

    let preRet: any = [], ret: any = [], postRet: any = []
    for (let plugin of plugins) {
        if (plugin.enforce) {
            plugin.enforce === 'pre' ? preRet.push(plugin) : postRet.push(plugin)
        } else {
            ret.push(plugin)
        }
    }

    return [...preRet, ...ret, ...postRet]
}

export function merge(plugins: ResolvedPlugin[]): Plugin {
    return {
        name: 'esbuild-plugin-merge',
        async setup(build) {

            const { esbuild: { build: builder } } = build
            proxyBuilder.initialOptions = build.initialOptions
            proxyBuilder.esbuild = build.esbuild
            plugins = sortPlugins(plugins)



            for (let plugin of plugins) {

                await plugin.setup(proxyBuilder as any)
            }


            for (let cb of proxyBuilder.startCb) {
                build.onStart(async () => {
                    return cb()
                })
            }
            for (let cb of proxyBuilder.endCb) {
                build.onEnd(async (param) => {
                    return cb(param)
                })
            }

            build.onResolve({ filter: /.*/ }, async (args) => {
                for (let item of proxyBuilder.resolveCb) {
                    if (item.filter.test(args.path)) {
                        let ret = await item.callback(args)
                        if (ret) return ret
                    }
                }
            })

            build.onLoad({ filter: /.*/ }, async (args) => {
                let loadRes: any
                if (fs.existsSync(args.path)) {
                    watchList.add(resolve(args.path))

                }
                for (let item of proxyBuilder.loadCb) {
                    if (item.filter.test(args.path)) {
                        loadRes = await item.callback(args)
                        if (loadRes) break
                    }
                }

                for (let item of proxyBuilder.transformCb) {
                    if (item.filter.test(args.path)) {
                        await item.callback(loadRes, args)
                    }
                }
                return loadRes as OnLoadResult
            })

            const rebuild = () => builder({
                ...build.initialOptions,
                watch: false,
            })


            build.onEnd(() => {

                if (build.initialOptions.watch) {
                    Object.keys(watchListRecord).forEach((id) => {
                        if (!watchList.has(id)) {
                            watchListRecord[id].close()
                            delete watchListRecord[id]
                        }
                    })
                    watchList.forEach((id) => {
                        if (!Object.keys(watchListRecord).includes(id)) {
                            watchListRecord[id] = chokidar.watch(id)
                            watchListRecord[id].on('change', async () => {
                                for (let cb of proxyBuilder.updateCb) {
                                    await cb(id, { event: 'update' })

                                } rebuild()
                            })
                            watchListRecord[id].on('unlink', async () => {
                                for (let cb of proxyBuilder.updateCb) {
                                    await cb(id, { event: 'delete' })

                                }
                                rebuild()
                            })
                        }
                    })
                }


            })
        }

    }


}

export * from './types'
