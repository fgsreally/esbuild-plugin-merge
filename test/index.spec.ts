import { build, Plugin } from 'esbuild'
import { merge, pre } from '../src'
import { test, expect, vi } from 'vitest'
import path from 'path'
import fs from 'fs'
import { ProPlugin } from '../src/types'

test('merge plugins order', async () => {

    const fn = vi.fn((id: string) => id)
    let initOptions: any
    let pluginA: ProPlugin = {
        name: 'pluginA',
        setup(build) {
            build.onStart(()=>{fn('startA')})
            
            build.onEnd(()=>{fn('endA')})

            initOptions = build.initialOptions
            build.onResolve({ filter: /\.css/, }, (args) => {
                fn('resolve-pluginA')
                return { path: path.join(args.resolveDir, args.path) }
            })
            build.onLoad({ filter: /\.css/, }, async (args) => {
                fn('load-pluginA')
                let text = await fs.promises.readFile(args.path, 'utf8')

                return {
                    contents: text,
                    loader: 'css'
                }
            })
            build.onTransForm({ filter: /\.css/ }, (args) => {
                fn('transform-pluginA')
                args.contents = 'console.log("pluginA")'

            })
            // build.onTransform()
        },
    }

    let pluginB: ProPlugin = {
        name: 'pluginB',
        setup(build) {
build.onStart(()=>{fn('startB')})
build.onEnd(()=>{fn('endB')})

            build.onResolve({ filter: /\.css/, }, (args) => {
                fn('resolve-pluginB')
                return { path: path.join(args.resolveDir, args.path) }
            })
            build.onLoad({ filter: /\.css/ }, async (args) => {
                fn('load-pluginB')

                return {
                    contents: 'console.log("pluginB")',
                    loader: 'js',
                }
            })

            build.onTransForm({ filter: /\.css/ }, () => {
                fn('transform-pluginB')

            })
        },
    }

    const ret = await build({
        format: 'esm',
        entryPoints: ['test/fixtures/main.ts'],
        write: false,
        plugins: [
            merge([pluginA, pre(pluginB)])
        ],
        bundle: true,
    })

    expect(initOptions).toMatchSnapshot()

    expect(fn).toHaveBeenCalledTimes(8)
    expect(fn).toHaveNthReturnedWith(3, 'resolve-pluginB')
    expect(fn).toHaveNthReturnedWith(4, 'load-pluginB')
    expect(fn).toHaveNthReturnedWith(5, 'transform-pluginB')
    expect(fn).toHaveNthReturnedWith(6, 'transform-pluginA')

    expect(ret.outputFiles[0].text).toMatchSnapshot()
})