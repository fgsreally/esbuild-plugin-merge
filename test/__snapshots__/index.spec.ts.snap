// Vitest Snapshot v1

exports[`merge plugins order 1`] = `
{
  "bundle": true,
  "entryPoints": [
    "test/fixtures/main.ts",
  ],
  "format": "esm",
  "plugins": [
    {
      "name": "esbuild-plugin-merge",
      "setup": [Function],
    },
  ],
  "write": false,
}
`;

exports[`merge plugins order 2`] = `
"// test/fixtures/test.ts
console.log(\\"test\\");

// test/fixtures/style.css
console.log(\\"pluginA\\");
"
`;
